README


This archive contains the distribution Test-DistManifest,
version 1.012:

  Author test that validates a package MANIFEST

This software is copyright (c) 2012 by Jonathan Yu <jawnsy@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
1.012     2012-04-22 13:45:02 America/Vancouver

          Now taking default MANIFEST.SKIP content directly from installed
          version of ExtUtils::Manifest, rather than inlining a copy (Karen
          Etheridge)

1.011     2011-04-24 12:43:21 America/Toronto

          (Previously released as 1.010-TRIAL)

          Migrate package to Dist::Zilla

          Drop a bunch of apparently-unnecessary dependencies. Now only
          support Perl 5.8.0+

1.009     2009-12-28

          Fixes tests so that the "default MANIFEST.SKIP" test is not run;
          this could fail when the default MANIFEST.SKIP provided by
          ExtUtils::Manifest doesn't cover special directories like debian/
          (Debian metadata)

1.008     2009-12-28

          Updated POD

1.007     2009-12-22

          Update skip info in t/circular (hopefully this fixes failures)

1.006     2009-12-21

          Update MANIFEST.SKIP according to new defaults provided by
          ExtUtils::Manifest

          Cleaned up some documentation issues

          Use the default MANIFEST.SKIP if ours can't be parsed

          Updated tests to match new behaviour

          Remove pass-through Build.PL

1.005     2009-10-12

          Fix the dist (don't overwrite the Changes file)

1.004     2009-10-11

          Changed my mail address to jawnsy@cpan.org

          Update MANIFEST.SKIP to include MYMETA.yml (prevents test failure)

1.003     2009-08-20

          Remove generated files from the repository

          Use a simple string version, per David Golden's article:
          http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/

          No longer use Perl::Critic to test

          Remove modules available in core since Perl 5.6 from requires

          Move CIRCULAR.SKIP and MANIFEST.EXTRA into t/

1.2.3     2009-07-25

          Added coverage information

          Rewrote LICENSE and README slightly

          Updated the checkmanifest.t example file and synopsis, it now
          checks Alias' flags (AUTOMATED_TESTING, RELEASE_TESTING)
          rather than TEST_AUTHOR

          No longer test Perl::Critic during AUTOMATED_TESTING, only
          only RELEASE_TESTING. Thanks to Elliot Shank for the
          discussion in a private mail.

          No more versioned dependency on version.pm

          Use a generic README, no longer produce one

          Changed critic.rc to ignore PodSpelling policy (fixes a bug
          in Debian, BTS#538115 -- when Pod::Spell and other
          dependencies are installed)

          Disable "one dot only" in filenames (portability test). This
          breaks compatibility with OpenVMS.

          Updated perlcriticrc with new preferences, removed "no critic"
          stuff for those in Test/DistManifest.pm

          Added copyleft statements to test files

          Minor documentation fix

          Small POD fix

1.2.2     2009-05-15

          Minor problem forgetting to bump one of the version numbers
          in META.yml

          Adjusted dependencies - changed minimum version dependency
          to Test::Builder 0.72 and version 0.74, which are core modules
          in Perl 5.10

1.2.1     2009-05-14

          Moved some small changes from MANIFEST.SKIP to CIRCULAR.SKIP

          This release is particularly important to get rid of Win32
          test failures, where `Build.bat' is created instead of `Build'

1.2.0     2009-05-14

          Updated t/03warn-only.t test with new semantics (always fail
          if parsing fails)

          Note that circular dependencies and parsing errors will
          always be considered fatal, regardless of the setting
          MANIFEST_WARN_ONLY

          Suppress refactoring requirement - even though manifest_ok
          is really complex, it doesn't make sense to exist in separate
          subs

          Added new feature for nonfatal errors (see updated POD)

1.1.4     2009-04-26

          Added .tmp to MANIFEST.SKIP

          Added Build.bat to MANIFEST.SKIP for Win32 builds...

1.1.3     2009-04-23

          Added Win32 compatibility code. Closes the Win32 compatibility
          bug (RT#44401)

          Updated Build.PL

          Added META.yml

          Changed to a passthrough Makefile instead of a classic one

          Updated version dependency for version.pm

          Removed use of Test::Prereq, because it needs to use CPAN
          to do its work

          Fix OS detection routines

          Modified the module so that it works portably (even with
          Win32, and other platforms File::Spec supports)

1.1.2     2009-03-14

          Updated prerequisite information in Build.PL and META.yml

          Add a more thorough description for the POD

          Fix dependencies in Build.PL and META.yml

          Clarified licensing restrictions

          CIRCULAR.SKIP became out of sync with the real MANIFEST.SKIP,
          and so was causing problems with debian/

1.1.1     2009-03-12

          Added "^debian/" as a mask for MANIFEST.SKIP. This is where
          Debian package control files go, and so should not be
          considered part of the package. Hopefully this won't break
          anything :-)

1.1.0     2009-03-12

          Modified Changelog name to "Changes" (instead of CHANGES)

          Updated Makefile.PL

          Added a test for circular dependencies (those that exist in
          both MANIFEST and MANIFEST.SKIP) -- thanks Apocalypse!

          Updated the POD and the README to reflect this repository's
          location

          Fixed version number in META.yml -> numeric "1.001000" style
          version numbering, to get around version.pm bugs

          Replaced PerlCritic's POD section to enforce 'LICENSE'
          instead of 'COPYRIGHT' (which is used by CPANTS/Kwalitee)

          Removed Test::NoWarnings from the Signature test, which
          throws warnings on purpose.

1.0       2008-12-25


lib/Test/DistManifest.pm
package Test::DistManifest;
{
  $Test::DistManifest::VERSION = '1.012';
}
# ABSTRACT: Author test that validates a package MANIFEST

use strict;
use warnings;
use Carp ();
use ExtUtils::Manifest;


# File management commands
use Cwd ();
use File::Spec; # Portability
use File::Spec::Unix; # To get UNIX-style paths
use File::Find (); # Traverse the filesystem tree

use Module::Manifest;
use Test::Builder;

my $test = Test::Builder->new;

my @EXPORTS = (
  'manifest_ok',
);

# These platforms were copied from File::Spec
my %platforms = (
  MacOS   => 1,
  MSWin32 => 1,
  os2     => 1,
  VMS     => 1,
  epoc    => 1,
  NetWare => 1,
  symbian => 1,
  dos     => 1,
  cygwin  => 1,
);

# Looking at other Test modules this seems to be an ad-hoc standard
sub import {
  my ($self, @plan) = @_;
  my $caller = caller;

  {
    no strict 'refs';
    for my $func (@EXPORTS) {
      *{$caller . '::' . $func} = \&{$func};
    }
  }

  $test->exported_to($caller);
  $test->plan(@plan);
  return;
}


sub manifest_ok {
  my $warn_only = $ENV{MANIFEST_WARN_ONLY} || 0;

  my $manifile = shift || 'MANIFEST';
  my $skipfile = shift || 'MANIFEST.SKIP';

  my $root = Cwd::getcwd(); # this is Build.PL's Cwd
  my $manifest = Module::Manifest->new;

  unless ($test->has_plan) {
    $test->plan(tests => 4);
  }

  # Try to parse the MANIFEST and MANIFEST.SKIP files
  eval {
    $manifest->open(manifest => $manifile);
  };
  if ($@) {
    $test->diag($!);
  }
  $test->ok(!$@, 'Parse MANIFEST or equivalent');

  eval {
    $manifest->open(skip => $skipfile);
  };
  if ($@) {
    $test->diag('Unable to parse MANIFEST.SKIP file:');
    $test->diag($!);
    $test->diag('Using default skip data from ExtUtils::Manifest ' . ExtUtils::Manifest->VERSION);

    open my $fh, '<', $ExtUtils::Manifest::DEFAULT_MSKIP
        or die "Cannot open $ExtUtils::Manifest::DEFAULT_MSKIP: $!";
    chomp(my @manifest_content = <$fh>);
    $manifest->parse( skip => \@manifest_content );
  }

  my @files;
  # Callback function called by File::Find
  my $closure = sub {
    # Trim off the package root to determine the relative path.
    my $path = File::Spec->abs2rel($File::Find::name, $root);

    # Portably deal with different OSes
    if ($platforms{$^O}) { # Check if we are on a non-Unix platform
      # Get path info from File::Spec, split apart
      my (undef, $dir, $file) = File::Spec->splitpath($path);
      my @dir = File::Spec->splitdir($dir);

      # Reconstruct the path in Unix-style
      $dir = File::Spec::Unix->catdir(@dir);
      $path = File::Spec::Unix->catpath(undef, $dir, $file);
    }

    # Test that the path is a file and then make sure it's not skipped
    if (-f $path && !$manifest->skipped($path)) {
      push @files, $path;
    }
    return;
  };

  # Traverse the directory recursively
  File::Find::find({
    wanted            => $closure,
    untaint           => 1,
    no_chdir          => 1,
  }, $root);

  # The two arrays have no duplicates. Thus we loop through them and
  # add the result to a hash.
  my %seen;
  # Allocate buckets for the hash
  keys(%seen) = 2 * scalar(@files);
  foreach my $path (@files, $manifest->files) {
    $seen{$path}++;
  }

  my $flag = 1;
  foreach my $path (@files) {
    # Skip the path if it was seen twice (the expected condition)
    next if ($seen{$path} == 2);

    # Oh no, we have files in @files not in $manifest->files
    if ($flag == 1) {
      $test->diag('Distribution files are missing in MANIFEST:');
      $flag = 0;
    }
    $test->diag($path);
  }
  $test->ok($warn_only || $flag, 'All files are listed in MANIFEST or ' .
    'skipped');

  # Reset the flag and test $manifest->files now
  $flag = 1;
  my @circular = (); # for detecting circular logic
  foreach my $path ($manifest->files) {
    # Skip the path if it was seen twice (the expected condition)
    next if ($seen{$path} == 2);

    # If the file should exist but is passed by MANIFEST.SKIP, we have
    # a strange circular logic condition.
    if ($manifest->skipped($path)) {
      push (@circular, $path);
      next;
    }

    # Oh no, we have files in $manifest->files not in @files
    if ($flag == 1) {
      $test->diag('MANIFEST lists the following missing files:');
      $flag = 0;
    }
    $test->diag($path);
  }
  $test->ok($warn_only || $flag, 'All files listed in MANIFEST exist ' .
    'on disk');

  # Test for circular dependencies
  $flag = (scalar @circular == 0) ? 1 : 0;
  if (not $flag) {
    $test->diag('MANIFEST and MANIFEST.SKIP have circular dependencies:');
    foreach my $path (@circular) {
      $test->diag($path);
    }
  }
  $test->ok($flag, 'No files are in both MANIFEST and MANIFEST.SKIP');

  return;
}


1;

__END__
=pod

=head1 NAME

Test::DistManifest - Author test that validates a package MANIFEST

=head1 VERSION

version 1.012

=head1 SYNOPSIS

  use Test::More;

  # This is the common idiom for author test modules like this, but see
  # the full example in examples/checkmanifest.t and, more importantly,
  # Adam Kennedy's article: http://use.perl.org/~Alias/journal/38822
  eval 'use Test::DistManifest';
  if ($@) {
    plan skip_all => 'Test::DistManifest required to test MANIFEST';
  }

  manifest_ok('MANIFEST', 'MANIFEST.SKIP'); # Default options

  manifest_ok(); # Functionally equivalent to above

=head1 DESCRIPTION

This module provides a simple method of testing that a MANIFEST matches the
distribution.

It tests three things:

=over

=item 1

Everything in B<MANIFEST> exists

=item 2

Everything in the package is listed in B<MANIFEST>, or subsequently matches
a regular expression mask in B<MANIFEST.SKIP>

=item 3

Nothing exists in B<MANIFEST> that also matches a mask in B<MANIFEST.SKIP>,
so as to avoid an unsatisfiable dependency conditions

=back

If there is no B<MANIFEST.SKIP> included in your distribution, this module
will replicate the toolchain behaviour of using the default system-wide
MANIFEST.SKIP file. To view the contents of this file, use the command:

  $ perldoc -m ExtUtils::MANIFEST.SKIP

=head1 EXPORTS

By default, this module exports the following functions:

=over

=item * manifest_ok

=back

=head1 FUNCTIONS

=head2 manifest_ok

  manifest_ok( $manifest, $skipfile )

This subroutine checks the manifest list contained in C<$manifest> by using
C<Module::Manifest> to determine the list of files and then checking for the
existence of all such files. Then, it checks if there are any files in the
distribution that were not specified in the C<$manifest> file but do not match
any regular expressions provided in the C<$skipfile> exclusion file.

If your MANIFEST file is generated by a module installation toolchain system
such as L<ExtUtils::MakeMaker>, L<Module::Build> or L<Module::Install>, then
you shouldn't have any problems with these files. It's just a helpful test
to remind you to update these files, using:

  $ make manifest # For ExtUtils::MakeMaker
  $ ./Build manifest # For Module::Build

=head2 Non-Fatal Errors

By default, errors in the B<MANIFEST> or B<MANIFEST.SKIP> files are treated
as fatal, which really is the purpose of using C<Test::DistManifest> as part
of your author test suite.

In some cases this is not desirable behaviour, such as with the Debian Perl
Group, which runs all tests - including author tests - as part of its module
packaging process. This wreaks havoc because Debian adds its control files
in C<debian/> downstream, and that directory or its files are generally not
in B<MANIFEST.SKIP>.

By setting the environment variable B<MANIFEST_WARN_ONLY> to a true value,
errors will be non-fatal - they show up as diagnostic messages only, but all
tests pass from the perspective of C<Test::Harness>.

This can be used in a test script as:

  $ENV{MANIFEST_WARN_ONLY} = 1;

or from other shell scripts as:

  export MANIFEST_WARN_ONLY=1

Note that parsing errors in B<MANIFEST> and circular dependencies will
always be considered fatal. The author is not aware of any cases where
other behaviour would be useful.

=head1 GUTS

This module internally plans 4 tests:

=over

=item 1

B<MANIFEST> can be parsed by C<Module::Manifest>

=item 2

Check which files exist in the distribution directory that do not match an
existing regular expression in B<MANIFEST.SKIP> and not listed in the
B<MANIFEST> file. These files should either be excluded from the test by
addition of a mask in MANIFEST.SKIP (in the case of temporary development
or test files) or should be included in the MANIFEST.

=item 3

Check which files are specified in B<MANIFEST> but do not exist on the disk.
This usually occurs when one deletes a test or similar script from the
distribution, or accidentally moves it.

=item 4

Check which files are specified in both B<MANIFEST> and B<MANIFEST.SKIP>.
This is clearly an unsatisfiable condition, since the file in question
cannot be expected to be included while also simultaneously ignored.

=back

If you want to run tests on multiple different MANIFEST files, you can
simply pass 'no_plan' to the import function, like so:

  use Test::DistManifest 'no_plan';

  # Multiple tests work properly now
  manifest_ok('MANIFEST', 'MANIFEST.SKIP');
  manifest_ok();
  manifest_ok('MANIFEST.OTHER', 'MANIFEST.SKIP');

I doubt this will be useful to users of this module. However, this is used
internally for testing and it might be helpful to you. You can also plan
more tests, but keep in mind that the idea of "3 internal tests" may change
in the future.

Example code:

  use Test::DistManifest tests => 5;
  manifest_ok(); # 4 tests
  ok(1, 'is 1 true?');

=head1 ACKNOWLEDGEMENTS

=over

=item *

Thanks to Adam Kennedy for developing L<Module::Manifest>, which provides
much of the core functionality for these tests.

=item *

Thanks to Apocalypse E<lt>apocal@cpan.orgE<gt>, for helping me track down
an obscure bug caused by circular dependencies: when files are expected by
MANIFEST but explictly skipped by MANIFEST.SKIP.

=back

=head1 SEE ALSO

L<Test::CheckManifest>, a module providing similar functionality

=head1 CAVEATS

=over

=item *

There is currently no way to test a MANIFEST/MANIFEST.SKIP without having the
files actually exist on disk. I am planning for this to change in the future.

=item *

This module has not been tested very thoroughly with Unicode.

=item *

This module does not produce any useful diagnostic messages in terms of how
to correct the situation. Hopefully this will be obvious for anybody using
the module; the emphasis should be on generating helpful error messages.

=back

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-DistManifest

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Jonathan Yu <jawnsy@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Jonathan Yu <jawnsy@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL
#!/usr/bin/perl

# Build.PL
#  Script to build and install this distribution
#
# $Id: Build.PL 10579 2009-12-22 01:16:57Z FREQUENCY@cpan.org $

use strict;
use warnings;

use Module::Build;

my $builder = Module::Build->new(
  module_name           => 'Test::DistManifest',
  license               => 'perl',
  dist_author           => 'Jonathan Yu <jawnsy@cpan.org>',
  dist_version_from     => 'lib/Test/DistManifest.pm',
  dynamic_config        => 0,
  create_readme         => 0,
  recursive_test_files  => 1,
  sign                  => 1,
  create_packlist       => 1,

  requires => {
    'perl'              => 5.006,

    # The tests are based on Test::More
    'Test::More'        => 0.62,
    'Test::Builder'     => 0.72,

    # Other modules
    'Module::Manifest'  => 0.07,
  },
  build_requires => {
    # User tests for good functionality
    'Test::Builder::Tester'   => 0,
    'Test::NoWarnings'        => 0.084,
  },
  recommends => {
  },
  conflicts => {
  },

  add_to_cleanup => [ 'Test-DistManifest-*' ],
  script_files => [],

  meta_merge => {
    resources => {
      # Custom resources (must begin with an uppercase letter)
      Ratings      => 'http://cpanratings.perl.org/d/Test-DistManifest',

      # Official keys (homepage, license, bugtracker)
      repository   => 'http://svn.ali.as/cpan/trunk/Test-DistManifest',
      bugtracker   => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-DistManifest',
      license      => 'http://dev.perl.org/licenses/',
    },
  },
);

$builder->create_build_script();


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Author test that validates a package MANIFEST",
  "AUTHOR" => "Jonathan Yu <jawnsy\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Test::More" => "0.62",
    "Test::NoWarnings" => "0.084"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Test-DistManifest",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::DistManifest",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Cwd" => 0,
    "ExtUtils::Manifest" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "Module::Manifest" => "0.07",
    "Test::Builder" => 0
  },
  "VERSION" => "1.012",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



